File: python/racored/kubo.py
from __future__ import annotations

import asyncio
import os
import shutil
from pathlib import Path

import httpx

from .config import data_directory


class KuboManager:
    """Starts the bundled IPFS Kubo node when one is not already available."""

    def __init__(self, api_url: str, gateway_url: str) -> None:
        self.api_url = api_url.rstrip("/")
        self.gateway_url = gateway_url.rstrip("/")
        self.process: asyncio.subprocess.Process | None = None
        self.executable = self._find_executable()
        self.repo = data_directory() / "ipfs"

    @staticmethod
    def _find_executable() -> str | None:
        explicit = os.getenv("RACORE_KUBO_PATH")
        candidates = [
            explicit,
            str(Path(__file__).resolve().parents[2] / "desktop" / "runtime" / "kubo" / "ipfs.exe"),
            shutil.which("ipfs"),
        ]
        return next((item for item in candidates if item and Path(item).exists()), None)

    async def _online(self) -> bool:
        try:
            async with httpx.AsyncClient(timeout=1.2) as client:
                response = await client.post(f"{self.api_url}/api/v0/id")
                return response.is_success
        except httpx.HTTPError:
            return False

    async def start(self) -> dict[str, object]:
        if await self._online():
            return {"managed": False, "online": True, "reason": "external-node"}
        if not self.executable:
            return {"managed": False, "online": False, "reason": "kubo-not-installed"}

        self.repo.mkdir(parents=True, exist_ok=True)
        env = {**os.environ, "IPFS_PATH": str(self.repo)}
        if not (self.repo / "config").exists():
            init = await asyncio.create_subprocess_exec(
                self.executable,
                "init",
                "--profile=lowpower",
                env=env,
                stdout=asyncio.subprocess.DEVNULL,
                stderr=asyncio.subprocess.DEVNULL,
            )
            if await init.wait() != 0:
                return {"managed": False, "online": False, "reason": "kubo-init-failed"}

        gateway_port = self.gateway_url.rsplit(":", 1)[-1]
        configure = await asyncio.create_subprocess_exec(
            self.executable,
            "config",
            "Addresses.Gateway",
            f"/ip4/127.0.0.1/tcp/{gateway_port}",
            env=env,
            stdout=asyncio.subprocess.DEVNULL,
            stderr=asyncio.subprocess.DEVNULL,
        )
        try:
            await asyncio.wait_for(configure.wait(), timeout=15)
        except asyncio.TimeoutError:
            configure.terminate()
            await configure.wait()
            return {"managed": False, "online": False, "reason": "kubo-config-timeout"}

        self.process = await asyncio.create_subprocess_exec(
            self.executable,
            "daemon",
            "--enable-gc",
            env=env,
            stdout=asyncio.subprocess.DEVNULL,
            stderr=asyncio.subprocess.DEVNULL,
        )
        for _ in range(40):
            if await self._online():
                return {"managed": True, "online": True, "pid": self.process.pid}
            if self.process.returncode is not None:
                break
            await asyncio.sleep(0.25)
        return {"managed": True, "online": False, "reason": "kubo-start-timeout"}

    async def stop(self) -> None:
        if self.process and self.process.returncode is None:
            self.process.terminate()
            try:
                await asyncio.wait_for(self.process.wait(), timeout=5)
            except asyncio.TimeoutError:
                self.process.kill()
                await self.process.wait()

File: god/internal/kubo/kubo.go
package kubo

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"time"
)

type Manager struct {
	apiURL     string
	gatewayURL string
	repoPath   string
	cmd        *exec.Cmd
	execPath   string
}

func New(apiURL, gatewayURL, dataDir string) *Manager {
	return &Manager{
		apiURL:     apiURL,
		gatewayURL: gatewayURL,
		repoPath:   filepath.Join(dataDir, "ipfs"),
	}
}

func (m *Manager) findExecutable() string {
	if v := os.Getenv("RACORE_KUBO_PATH"); v != "" {
		if _, err := os.Stat(v); err == nil {
			return v
		}
	}
	if p, err := exec.LookPath("ipfs"); err == nil {
		return p
	}
	return ""
}

func (m *Manager) isOnline(ctx context.Context) bool {
	req, err := http.NewRequestWithContext(ctx, "POST", m.apiURL+"/api/v0/id", nil)
	if err != nil {
		return false
	}
	client := &http.Client{Timeout: 1 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return false
	}
	resp.Body.Close()
	return resp.StatusCode == 200
}

func (m *Manager) Start(ctx context.Context) (map[string]any, error) {
	if m.isOnline(ctx) {
		return map[string]any{"managed": false, "online": true, "reason": "external-node"}, nil
	}

	m.execPath = m.findExecutable()
	if m.execPath == "" {
		return map[string]any{"managed": false, "online": false, "reason": "kubo-not-installed"}, nil
	}

	os.MkdirAll(m.repoPath, 0700)
	env := append(os.Environ(), "IPFS_PATH="+m.repoPath)

	if _, err := os.Stat(filepath.Join(m.repoPath, "config")); err != nil {
		initCmd := exec.CommandContext(ctx, m.execPath, "init", "--profile=lowpower")
		initCmd.Env = env
		if out, err := initCmd.CombinedOutput(); err != nil {
			return map[string]any{"managed": false, "online": false, "reason": fmt.Sprintf("kubo-init-failed: %s", string(out))}, nil
		}
	}

	gatewayPort := "8180"
	if n, err := fmt.Sscanf(m.gatewayURL, "%*s%d", &gatewayPort); err != nil && n != 1 {
		gatewayPort = "8180"
	}

	configCmd := exec.CommandContext(ctx, m.execPath, "config", "Addresses.Gateway", fmt.Sprintf("/ip4/127.0.0.1/tcp/%s", gatewayPort))
	configCmd.Env = env
	configCmd.Run()

	m.cmd = exec.CommandContext(ctx, m.execPath, "daemon", "--enable-gc")
	m.cmd.Env = env
	m.cmd.Stdout = nil
	m.cmd.Stderr = nil

	if err := m.cmd.Start(); err != nil {
		return map[string]any{"managed": true, "online": false, "reason": fmt.Sprintf("kubo-daemon-start: %v", err)}, nil
	}

	for range 40 {
		if m.isOnline(ctx) {
			return map[string]any{"managed": true, "online": true, "pid": m.cmd.Process.Pid}, nil
		}
		if m.cmd.ProcessState != nil && m.cmd.ProcessState.Exited() {
			break
		}
		time.Sleep(250 * time.Millisecond)
	}

	m.Stop()
	return map[string]any{"managed": true, "online": false, "reason": "kubo-start-timeout"}, nil
}

func (m *Manager) Stop() {
	if m.cmd != nil && m.cmd.Process != nil {
		m.cmd.Process.Signal(os.Interrupt)
		done := make(chan any, 1)
		go func() {
			m.cmd.Wait()
			done <- nil
		}()
		select {
		case <-done:
		case <-time.After(5 * time.Second):
			m.cmd.Process.Kill()
		}
	}
}
